asv_bench/benchmarks/utils.py

```diff
@@ -217,6 +217,89 @@ def gen_str_int_data(nrows: int, ncols: int, rand_low: int, rand_high: int) -> d
     return data
 
 
+def gen_true_false_data(nrows: int, ncols: int, true_false_values: list):
+    """
+    Generate data with "true" and "false" values with caching.
+
+    The generated data are saved in the dictionary and on a subsequent call,
+    if the keys match, saved data will be returned. Therefore, we need
+    to carefully monitor the changing of saved data and make its copy if needed.
+
+    Parameters
+    ----------
+    nrows : int
+        Number of rows.
+    ncols : int
+        Number of columns.
+    true_false_values : list
+        List with strings that represent "true" and "false" values.
+
+    Returns
+    -------
+    dict
+        Number of keys - `ncols`, each of them store np.ndarray of `nrows` length.
+    """
+    cache_key = ("true_false", nrows, ncols, "_".join(true_false_values))
+    if cache_key in data_cache:
+        return data_cache[cache_key]
+
+    logging.info(
+        "Generating true_false data {} rows and {} columns with values {}".format(
+            nrows, ncols, true_false_values
+        )
+    )
+    data = {
+        "tf_col{}".format(i): random_state.choice(true_false_values, size=(nrows))
+        for i in range(ncols)
+    }
+    data_cache[cache_key] = weakdict(data)
+    return data
+
+
+def gen_true_false_int_data(nrows, ncols, rand_low, rand_high):
+    """
+    Generate int data and string data "true" and "false" values with caching.
+
+    The generated data are saved in the dictionary and on a subsequent call,
+    if the keys match, saved data will be returned. Therefore, we need
+    to carefully monitor the changing of saved data and make its copy if needed.
+
+    Parameters
+    ----------
+    nrows : int
+        Number of rows.
+    ncols : int
+        Number of columns.
+    rand_low : int
+        Low bound for random generator.
+    rand_high : int
+        High bound for random generator.
+
+    Returns
+    -------
+    dict
+        Number of keys - `ncols`, each of them store np.ndarray of `nrows` length.
+        One half of the columns with integer values, another half - with "true" and
+        "false" string values.
+    """
+    cache_key = ("true_false_int", nrows, ncols, rand_low, rand_high)
+    if cache_key in data_cache:
+        return data_cache[cache_key]
+
+    logging.info(
+        "Generating true_false_int data {} rows and {} columns [{}-{}]".format(
+            nrows, ncols, rand_low, rand_high
+        )
+    )
+    data = gen_int_data(nrows // 2, ncols // 2, rand_low, rand_high).copy()
+    data_true_false = gen_true_false_data(
+        nrows - nrows // 2, ncols - ncols // 2, ["Yes", "true", "No", "false"]
+    ).copy()
+    data.update(data_true_false)
+    data_cache[cache_key] = weakdict(data)
+    return data
```

Comment:
This is a rather specific case, it will not be used widely, so I suggest not to divide the functionality into two functions and leave everything in `gen_true_false_int_data`.

Also, given that caching is a fairly general mechanism, then maybe we should move caching into `gen_data` function (to reduce copy-paste). What do you think?

Reply:
Ok, all true/false data generation was moved to the `gen_true_false_int_data` function.

I think it is good idea, implemented in the last commit.